tests/components/canary/test_config_flow.py

```diff
@@ -0,0 +1,121 @@
+    await async_setup_component(hass, "persistent_notification", {})
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["errors"] == {}
+
+    with _patch_async_setup() as mock_setup, _patch_async_setup_entry() as mock_setup_entry:
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            USER_INPUT,
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == RESULT_TYPE_CREATE_ENTRY
+    assert result["title"] == "test-username"
+    assert result["data"] == {**USER_INPUT, CONF_TIMEOUT: DEFAULT_TIMEOUT}
+
+    assert len(mock_setup.mock_calls) == 1
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_user_form_cannot_connect(hass, canary_config_flow):
+    """Test we handle errors that should trigger the cannot connect error."""
+    canary_config_flow.side_effect = HTTPError()
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        USER_INPUT,
+    )
+
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["errors"] == {"base": "cannot_connect"}
+
+    canary_config_flow.side_effect = ConnectTimeout()
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        USER_INPUT,
+    )
+
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["errors"] == {"base": "cannot_connect"}
+
+
+async def test_user_form_unexpected_exception(hass, canary_config_flow):
+    """Test we handle unexpected exception."""
+    canary_config_flow.side_effect = Exception()
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        USER_INPUT,
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "unknown"
+
+
+async def test_user_form_single_instance_allowed(hass, canary_config_flow):
+    """Test that configuring more than one instance is rejected."""
+    await init_integration(hass, skip_entry_setup=True)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_USER},
+        data=USER_INPUT,
+    )
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "single_instance_allowed"
+
+
+async def test_options_flow(hass):
+    """Test updating options."""
+    entry = await init_integration(hass, skip_entry_setup=True)
+    assert entry.options[CONF_FFMPEG_ARGUMENTS] == DEFAULT_FFMPEG_ARGUMENTS
+    assert entry.options[CONF_TIMEOUT] == DEFAULT_TIMEOUT
+
+    result = await hass.config_entries.options.async_init(entry.entry_id)
```

Comment:
The config entry should be setup before starting the options flow. The test might fail if the config flow module isn't loaded, otherwise. This can happen on CI where we split tests on different runners.

Reply:
You can test this locally by just running this test:
```
(home-assistant) martin@x:~/Dev/home-assistant/home-assistant$ pytest tests/components/canary/test_config_flow.py -k test_options_flow
Test session starts (platform: linux, Python 3.7.9, pytest 6.0.2, pytest-sugar 0.9.4)
rootdir: /home/martin/Dev/home-assistant/home-assistant, configfile: setup.cfg
plugins: sugar-0.9.4, mock-3.3.1, xdist-2.1.0, requests-mock-1.8.0, timeout-1.4.2, aiohttp-0.3.0, forked-1.3.0, rerunfailures-9.1, test-groups-1.0.3, cov-2.10.1
collecting ... 

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_options_flow[pyloop] ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

hass = <homeassistant.core.HomeAssistant object at 0x7fc1a90020d0>

    async def test_options_flow(hass):
        """Test updating options."""
        entry = await init_integration(hass, skip_entry_setup=True)
        assert entry.options[CONF_FFMPEG_ARGUMENTS] == DEFAULT_FFMPEG_ARGUMENTS
        assert entry.options[CONF_TIMEOUT] == DEFAULT_TIMEOUT
    
>       result = await hass.config_entries.options.async_init(entry.entry_id)

tests/components/canary/test_config_flow.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
homeassistant/data_entry_flow.py:120: in async_init
    flow = await self.async_create_flow(handler, context=context, data=data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <homeassistant.config_entries.OptionsFlowManager object at 0x7fc1add53710>, handler_key = '7ba0b819e1854f959829dc23f08d8e10'

    async def async_create_flow(
        self,
        handler_key: Any,
        *,
        context: Optional[Dict[str, Any]] = None,
        data: Optional[Dict[str, Any]] = None,
    ) -> "OptionsFlow":
        """Create an options flow for a config entry.
    
        Entry_id and flow.handler is the same thing to map entry with flow.
        """
        entry = self.hass.config_entries.async_get_entry(handler_key)
        if entry is None:
            raise UnknownEntry(handler_key)
    
        if entry.domain not in HANDLERS:
>           raise data_entry_flow.UnknownHandler
E           homeassistant.data_entry_flow.UnknownHandler

homeassistant/config_entries.py:1028: UnknownHandler
--------------------------------------------------------------------------------- Captured stderr setup ----------------------------------------------------------------------------------
DEBUG:asyncio:Using selector: EpollSelector
----------------------------------------------------------------------------------- Captured log setup -----------------------------------------------------------------------------------
DEBUG    asyncio:selector_events.py:58 Using selector: EpollSelector

 tests/components/canary/test_config_flow.py ⨯                                                                                                                             100% ██████████
================================================================================ short test summary info =================================================================================
FAILED tests/components/canary/test_config_flow.py::test_options_flow[pyloop] - homeassistant.data_entry_flow.UnknownHandler

Results (0.28s):
       1 failed
         - tests/components/canary/test_config_flow.py:104 test_options_flow[pyloop]
       4 deselected
```

Reply:
so would you suggest setting up but patching the platforms from running their setups?

Reply:
Depending on what we need to access in the options flow, we can patch the integration or platforms `async_setup_entry` or the `PLATFORMS` constant.